### PR TITLE
Rules implementation

### PR DESCRIPTION
**Pull request by Buckram123**
- [x] TODO: change `Cargo.toml` so it points to main branch (after [this](https://github.com/CronCats/cw-croncat/pull/104) gets merged)

### CODE REVIEW

File: croncat/src/grpc.rs [hunk 1]
```diff
@@ -25,7 +28,7 @@ use crate::logging::info;
 ///
 /// Create message and query clients for interacting with the chain.
 ///
-#[no_coverage]
+//#[no_coverage]
```
Comment (mikedotexe) [inline croncat/src/grpc.rs]:
```suggestion
```

File: croncat/src/grpc.rs [hunk 2]
```diff
@@ -126,12 +126,61 @@ impl GrpcSigner {
...
+    pub async fn fetch_rules(&self) -> Result<Vec<TaskWithRulesResponse>, Report> {
+        let mut tasks_with_rules = Vec::new();
+        let mut start_index = 0;
+        let limit = 20;
+        loop {
+            let current_iteration = self
+                .query_get_tasks_with_rules(Some(start_index), Some(limit))
+                .await?;
+            let last_iteration = current_iteration.len() < limit as usize;
+            tasks_with_rules.extend(current_iteration);
```
Comment (mikedotexe) [inline croncat/src/grpc.rs:159]: Awesome. AwesmWasm

File: croncat/src/lib.rs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 //!
 
 // Features
-#![feature(no_coverage)]
+//#![feature(no_coverage)]
```
Comment (mikedotexe) [inline croncat/src/lib.rs]:
```suggestion
```

File: croncat/src/streams/tasks.rs [hunk 4]
```diff
@@ -26,14 +28,20 @@ pub async fn tasks_loop(
     let block_consumer_stream = tokio::task::spawn(async move {
         while let Ok(_block) = block_stream_rx.recv().await {
             let locked_status = block_status.lock().await;
-            if *locked_status == AgentStatus::Active {
+            let is_active = *locked_status == AgentStatus::Active;
+            // unlocking it ASAP
+            std::mem::drop(locked_status);
```
Comment (mikedotexe) [inline croncat/src/streams/tasks.rs:33]: Can you explain what's happening with this `drop`? I'm a little confused

Comment (mikedotexe) [reply croncat/src/streams/tasks.rs:33]: actually cc'ing @SeedyROM who might find this interesting

Comment (SeedyROM) [reply croncat/src/streams/tasks.rs:33]: @mikedotexe 

The drop is to free the lock before the end of the function scope. Locks drop when out of scope but sometimes you need to release a lock before that. Also `std::mem` isn't the only way to call drop. You could do `.drop()` as well.

File: rust-toolchain [hunk 5]
```diff
@@ -1 +1 @@
-nightly
+stable
```
Comment (mikedotexe) [inline rust-toolchain:1]: ❤️

**Review by mikedotexe — APPROVED** [review]

File: croncat/src/streams/tasks.rs [hunk 6]
```diff
@@ -50,36 +58,54 @@ pub async fn tasks_loop(
...
-                .map_or(false, |ag| ag.status == AgentStatus::Active);
-            if agent_active {
-                let tasks = signer.get_agent_tasks(account_addr).await.unwrap();
-                if tasks.is_some() {
-                    if let Ok(proxy_call_res) = signer.proxy_call().await {
-                        info!("Finished task: {}", proxy_call_res.log);
-                    } else {
-                        warn!("Something went wrong during proxy_call");
+        while let Ok(block) = block_stream_rx.recv().await {
+            let tasks_with_rules = signer.fetch_rules().await.expect("Failed to fetch rules");
```
Comment (Buckram123) [inline croncat/src/streams/tasks.rs:69]: Don't like fetching rules every block. But there is a chance that someone completed the task, not sure what's the best thing to do here. Still good enough for MVP IMO

**Review by SeedyROM — APPROVED** [review]
LGTM :+1:

**Pull request merged**